### ISSUE TITLE
the astropy unit conversion causes a 10^-8 error in calculation

### ISSUE THREAD

**Issue by Firestar-Reimu** [astropy/astropy#17168]
### Description

```python
import numpy as np
from astropy.stats import sigma_clip
from sklearn.cluster import DBSCAN
import scipy.sparse as sp
from astropy.io import fits
from tqdm import trange
from astropy.wcs import WCS
from spectral_cube import SpectralCube as sc
import astropy.units as u

np.seterr(divide="ignore", invalid="ignore")
np.set_printoptions(precision=12, suppress=True)

file = "HVC_test.fits"
hdul = fits.open(file)
hdr = hdul[0].header
wcs = WCS(hdr)
data = hdul[0].data
data = np.nan_to_num(data)
print("data.shape =", data.shape)

delta_v = abs(hdr['CDELT3']) / 1000
print(delta_v)

cube = sc.read(hdul)
cube = cube.with_spectral_unit(u.km / u.s)

filtered_data = sigma_clip(data, sigma=3, maxiters=10, masked=True)
mask = filtered_data.mask
index = np.transpose(np.nonzero(mask)).astype(int)

db = DBSCAN(eps=np.sqrt(3), min_samples=18, n_jobs=-1).fit(index)
labels = db.labels_
unique_labels = set(labels)
n_clusters = len(unique_labels) - (1 if -1 in labels else 0)

# use numpy array to calculate moment 0

def moment_0_np(data, index, labels, n_clusters):
    moment_0_cube = np.empty(0)
    for i in trange(n_clusters):
        vyx = index[labels == i]  # the vyx coord of every point in cluster_i
        bool_array = np.zeros(data.shape, dtype=np.bool)
        bool_array[vyx[:,0], vyx[:,1], vyx[:,2]] = True  # set True for every point in cluster_i
        moment_0_coo = sp.coo_array(np.nansum(bool_array * data * delta_v, axis=0))
    moment_0_cube = np.hstack((moment_0_cube, moment_0_coo))
    return moment_0_cube

moment_0_cube_np = moment_0_np(data, index, labels, n_clusters)
print(moment_0_cube_np[0])

# use spectral cube to calculate moment 0

def moment_0_sc(data, index, labels, n_clusters):
    moment_0_cube = np.empty(0)
    for i in trange(n_clusters):
        vyx = index[labels == i]  # the vyx coord of every point in cluster_i
        bool_array = np.zeros(data.shape, dtype=np.bool)
        bool_array[vyx[:,0], vyx[:,1], vyx[:,2]] = True  # set True for every point in cluster_i
        bool_cube = sc(bool_array * data, wcs=wcs).with_spectral_unit(u.km / u.s)
        moment_0_coo = sp.coo_array(bool_cube.moment(order=0))
    moment_0_cube = np.hstack((moment_0_cube, moment_0_coo))
    return moment_0_cube

moment_0_cube_sc = moment_0_sc(data, index, labels, n_clusters)
print(moment_0_cube_sc[0])

print(moment_0_cube_np[0] - moment_0_cube_sc[0])
```

The data used in this links here: https://cowtransfer.com/s/abeea1c9dbed43

If you run my code you can see that:

```
(base) [firestar@Archlinux Desktop]$ python test.py
data.shape = (447, 403, 196)
0.201283594733
100%|█████████████████████████████████████| 617/617 [01:12<00:00,  8.52it/s]
<COOrdinate sparse array of dtype 'float32'
        with 21 stored elements and shape (403, 196)>
  Coords        Values
  (206, 138)    0.059925541281700134
  (206, 139)    0.11611424386501312
  (206, 140)    0.11767134070396423
  (206, 141)    0.13588958978652954
  (206, 142)    0.08509054780006409
  (207, 138)    0.20172874629497528
  (207, 139)    0.20145222544670105
  (207, 140)    0.20981991291046143
  (207, 141)    0.21626465022563934
  (207, 142)    0.21031689643859863
  (208, 138)    0.22760865092277527
  (208, 139)    0.23064976930618286
  (208, 140)    0.21912509202957153
  (208, 141)    0.1933283805847168
  (208, 142)    0.0700044184923172
  (209, 138)    0.0627964660525322
  (209, 139)    0.24036796391010284
  (209, 140)    0.2356765866279602
  (209, 141)    0.18532486259937286
  (210, 139)    0.052889805287122726
  (210, 140)    0.20767921209335327
100%|█████████████████████████████████████| 617/617 [02:21<00:00,  4.37it/s]
<COOrdinate sparse array of dtype 'float64'
        with 21 stored elements and shape (403, 196)>
  Coords        Values
  (206, 138)    0.059925542983726376
  (206, 139)    0.1161142442142389
  (206, 140)    0.11767134959261719
  (206, 141)    0.13588960189414312
  (206, 142)    0.08509055135147338
  (207, 138)    0.20172875364022103
  (207, 139)    0.201452224711047
  (207, 140)    0.2098199032464072
  (207, 141)    0.2162646565481282
  (207, 142)    0.2103169090783289
  (208, 138)    0.22760865929492866
  (208, 139)    0.23064976366833145
  (208, 140)    0.21912509151529078
  (208, 141)    0.1933284000847469
  (208, 142)    0.0700044219803887
  (209, 138)    0.06279646970968814
  (209, 139)    0.24036797570899443
  (209, 140)    0.23567659386020412
  (209, 141)    0.18532486776216334
  (210, 139)    0.05288980792478459
  (210, 140)    0.20767923052697496
<Compressed Sparse Row sparse array of dtype 'float64'
        with 21 stored elements and shape (403, 196)>
  Coords        Values
  (206, 138)    -1.7020262421674204e-09
  (206, 139)    -3.4922577329954407e-10
  (206, 140)    -8.888652952565579e-09
  (206, 141)    -1.2107613578393739e-08
  (206, 142)    -3.5514092916466566e-09
  (207, 138)    -7.345245744971862e-09
  (207, 139)    7.356540376868281e-10
  (207, 140)    9.66405422353489e-09
  (207, 141)    -6.322488865784237e-09
  (207, 142)    -1.2639730262442583e-08
  (208, 138)    -8.372153387226078e-09
  (208, 139)    5.6378514079469255e-09
  (208, 140)    5.14280756869212e-10
  (208, 141)    -1.9500030090657816e-08
  (208, 142)    -3.4880714983032135e-09
  (209, 138)    -3.657155939196244e-09
  (209, 139)    -1.1798891585979021e-08
  (209, 140)    -7.23224391485644e-09
  (209, 141)    -5.162790472912349e-09
  (210, 139)    -2.637661866700114e-09
  (210, 140)    -1.8433621684632584e-08
[ble: elapsed 224.207s (CPU 105.0%)] python test.py
```

So that there is a 10^-8 error between numpy and spectral-cube

### Expected behavior

**Then I changed the `delta_v = abs(hdr['CDELT3']) / 1000` to `delta_v = (abs(hdr['CDELT3']) * u.m / u.s).to(u.km / u.s).value`**

the 2 results become the same:

```
(base) [firestar@Archlinux Desktop]$ python test.py
data.shape = (447, 403, 196)
0.201283594733
100%|█████████████████████████████████████| 617/617 [01:53<00:00,  5.45it/s]
<COOrdinate sparse array of dtype 'float64'
        with 21 stored elements and shape (403, 196)>
  Coords        Values
  (206, 138)    0.059925542983726376
  (206, 139)    0.1161142442142389
  (206, 140)    0.11767134959261719
  (206, 141)    0.13588960189414312
  (206, 142)    0.08509055135147338
  (207, 138)    0.20172875364022103
  (207, 139)    0.201452224711047
  (207, 140)    0.2098199032464072
  (207, 141)    0.2162646565481282
  (207, 142)    0.2103169090783289
  (208, 138)    0.22760865929492866
  (208, 139)    0.23064976366833145
  (208, 140)    0.21912509151529078
  (208, 141)    0.1933284000847469
  (208, 142)    0.0700044219803887
  (209, 138)    0.06279646970968814
  (209, 139)    0.24036797570899443
  (209, 140)    0.23567659386020412
  (209, 141)    0.18532486776216334
  (210, 139)    0.05288980792478459
  (210, 140)    0.20767923052697496
100%|█████████████████████████████████████| 617/617 [02:29<00:00,  4.12it/s]
<COOrdinate sparse array of dtype 'float64'
        with 21 stored elements and shape (403, 196)>
  Coords        Values
  (206, 138)    0.059925542983726376
  (206, 139)    0.1161142442142389
  (206, 140)    0.11767134959261719
  (206, 141)    0.13588960189414312
  (206, 142)    0.08509055135147338
  (207, 138)    0.20172875364022103
  (207, 139)    0.201452224711047
  (207, 140)    0.2098199032464072
  (207, 141)    0.2162646565481282
  (207, 142)    0.2103169090783289
  (208, 138)    0.22760865929492866
  (208, 139)    0.23064976366833145
  (208, 140)    0.21912509151529078
  (208, 141)    0.1933284000847469
  (208, 142)    0.0700044219803887
  (209, 138)    0.06279646970968814
  (209, 139)    0.24036797570899443
  (209, 140)    0.23567659386020412
  (209, 141)    0.18532486776216334
  (210, 139)    0.05288980792478459
  (210, 140)    0.20767923052697496
<Compressed Sparse Row sparse array of dtype 'float64'
        with 0 stored elements and shape (403, 196)>
[ble: elapsed 274.153s (CPU 105.6%)] python test.py
```

**Although `abs(hdr['CDELT3']) / 1000 == (abs(hdr['CDELT3']) * u.m / u.s).to(u.km / u.s).value` return `np.True_`**

### Versions

Linux-6.11.2-arch1-1-x86_64-with-glibc2.40
Python 3.12.6 | packaged by conda-forge | (main, Sep 22 2024, 14:16:49) [GCC 13.3.0]
astropy 6.1.4
Numpy 2.0.2
pyerfa 2.0.1.4
Scipy 1.14.1
Matplotlib 3.9.2

**Comment by pllim** [comment]
> So that there is a 10^-8 error between numpy and spectral-cube

Are you sure this is an astropy bug and not a bug over at https://github.com/radio-astro-tools/spectral-cube ? cc @keflavich

**Comment by Firestar-Reimu** [comment]
> Then I changed the delta_v = abs(hdr['CDELT3']) / 1000 to delta_v = (abs(hdr['CDELT3']) * u.m / u.s).to(u.km / u.s).value

I think this is because the unit conversion.

**Comment by keflavich** [comment]
I can't quite follow what's going on here.  A 10^-8 error is quite often at the limit of numerical precision.  In the original bug report, I didn't see evidence that the numbers were identical in the second case.

If this is an error at the 10^-8 level (i.e., fractional error 10^-8), I would not consider this a bug, but it may be a consequence of converting between m/s and km/s dropping some precision when using float32.

**Comment by Firestar-Reimu** [comment]
> I didn't see evidence that the numbers were identical in the second case.

you can check digits one by one and you see that they are all the same

---

**The big problem is, which is considered more accurate? `x / 1000` or `x m/s --> km/s`?**

**Comment by ayshih** [comment]
Okay, there are two aspect to this, one straightforward and one surprising, both related to the bit depth of floats.  As @keflavich said, when working with 32-bit floats, the numerical precision will be limited to ~1e-8.

The straightforward aspect is that the first computation using `delta_v = abs(hdr['CDELT3']) / 1000` results in an array of 32-bit floats (`COOrdinate sparse array of dtype 'float32'`), while using `delta_v = (abs(hdr['CDELT3']) * u.m / u.s).to(u.km / u.s).value` results in an array of 64-bit floats (`COOrdinate sparse array of dtype 'float64'`).  For an accurate/consistent calculation, you simply need to make sure that your float array is 64 bits before doing any summing.  One way could be to convert the original data array to avoid any misunderstanding:
```
data = data.astype(np.float64)
```

The surprising aspect is _why_ these two expressions for `delta_v` lead to different behavior, and it's a "problem" with NumPy.  `delta_v` is a 64-bit float in both cases, but in the first case it is native Python `float` and in the second case it is a NumPy `float64`.  The problem then is triggered by the `data * delta_v` product later, where data is an array of `np.float32`

* a `np.float32` times a Python `float` stays a `np.float32`
* a `np.float32` times a Python `np.float64` becomes a `np.float64`

In other words, the first expression would behave how you want if you simply cast `delta_v` to be a NumPy 64-bit float rather than a Python float:
```
delta_v = np.float64(abs(hdr['CDELT3']) / 1000)
```

**Comment by ayshih** [comment]
It's worth discussing whether such header keyword values should be stored as NumPy floats rather than Python floats to avoid such computational surprises

**Comment by keflavich** [comment]
> a np.float32 times a Python float stays a np.float32

Whoa, I'm surprised at that!  I thought numpy operations alway promoted to the higher-precision value.  

 
I've encountered a lot of errors around weirdly-handled header values - since they're strings natively (headers are strings), it's possible to introduce weird rounding issues.  Some software (e.g., CARTA) only works with square pixels, and if there are differences in the n'th digit of the CDELT, it fails to load them.

I think you're right that we should read these values as np floats to avoid this issue.  I'm not sure it really relates to the side issue I reported, but that's a note that this can matter.

**Comment by mhvk** [comment]
>   a np.float32 times a Python float stays a np.float32

See [NEP 50](https://numpy.org/neps/nep-0050-scalar-promotion.html). In general this is simply tricky, as one can easily think of "very logical" examples where promotion should or should not happen. The decision to support weak promotion for python scalars is driven mostly by wanting `some_int_array + 1` to not automatically promote to more bytes (which would be very bad if `some_int_array` was large and had `dtype="i1"`). Similarly, `data = float32_array / 1000.0` will stay at `float32`, which seems *generally* what would be desired.

> It's worth discussing whether such header keyword values should be stored as NumPy floats rather than Python floats to avoid such computational surprises

I'm not sure that would be best - it feels like guessing what they are needed for. In the present example, if `data` were a really large array, would it necessarily be good if the size doubled? It would seem to depend on what the next operation is.

I think the surprise here is mostly that `Quantity(float)` does convert the float to a (view of a) `np.float64` array internally, and `value` keeps that dtype. Perhaps `Quantity` should remember its input in this case, although that also seems rather complicated... a `QuantityFloat` may be more appropriate...

**Comment by ayshih** [comment]
> > It's worth discussing whether such header keyword values should be stored as NumPy floats rather than Python floats to avoid such computational surprises
> 
> I'm not sure that would be best - it feels like guessing what they are needed for. In the present example, if `data` were a really large array, would it necessarily be good if the size doubled? It would seem to depend on what the next operation is.

To be clear, I was referring to the fact that although the bit depth of `hdul[0].data` is defined in the file (hence `np.float32`), header keywords such as `hdul[0].header[CDELT3]` are (understandably) typed as a general Python `float`.  So, even if the latter is making use of 64 bits, multiplying it against a `np.float32` then degrades the precision.

**Comment by mhvk** [comment]
> So, even if the latter is making use of 64 bits, multiplying it against a `np.float32` then degrades the precision.

Yes, that's true, but I think in the general case that is as desired: one generally does not want the array to double in size to keep precision in the factor one multiplies with; after all the array has less precision itself. But the general case often does not hold in specific ones, and it is not helpful that a `Quantity` scalar, like a numpy one, behaves differently.

Really, this is not a simple problem, and any solution one proposes is going to be bad for some use cases (as discussed at some length in NEP 50).

**Comment by pllim** [comment]
So nothing much we can do here? Should we close as won't fix?
